client/zanata-client-commands/src/main/java/org/zanata/client/commands/PutUserCommand.java

```diff
@@ -21,22 +21,46 @@ public PutUserCommand(PutUserOptions opts) {
     }
 
     public void run() throws Exception {
-        Account account = new Account();
-        account.setEmail(getOpts().getUserEmail());
-        account.setName(getOpts().getUserName());
-        account.setUsername(getOpts().getUserUsername());
-        account.setPasswordHash(getOpts().getUserPasswordHash());
-        account.setApiKey(getOpts().getUserKey());
+
+        String userUsername = getOpts().getUserUsername();
+        String passwordHash = getOpts().getUserPasswordHash();
+        String userEmail = getOpts().getUserEmail();
+
+        Account account = getClientFactory().getAccountClient().get(userUsername);
+
+        if (account == null) {
+            log.info("Creating new account {}", userUsername);
+            account = new Account();
+            if (userEmail == null || userUsername == null) {
+                throw new RuntimeException("New user's name and email must be specified");
+            } else if (passwordHash == null || StringUtils.isBlank(passwordHash)) {
+                log.warn("Unusable passwordHash set, user will need to use the Forgot Password feature");
+                account.setPasswordHash("NULL");
+            }
+            account.setUsername(userUsername);
+        } else {
+            log.info("Updating account {}", account.getUsername());
+        }
+
+        account.setEmail(Optional.ofNullable(userEmail).orElse(account.getEmail()));
+        account.setName(Optional.ofNullable(getOpts().getUserName()).orElse(account.getName()));
+        account.setApiKey(Optional.ofNullable(getOpts().getUserKey()).orElse(account.getApiKey()));
+        account.setPasswordHash(Optional.ofNullable(passwordHash).orElse(account.getPasswordHash()));
```

Comment:
Using Optional for these leads to ugly code. You'd be better off with ternary operators (or another language, but that's another problem).

eg

    account.setPasswordHash(passwordHash != null ? passwordHash : account.getPasswordHash());

EDIT: or try com.google.common.base.MoreObjects#firstNonNull